### PR TITLE
Per-connection CatalogCache prototype

### PR DESCRIPTION
**Pull request by mbutrovich**
Taken from Doxygen for `CatalogCache` object:

>  Simple cache for `DatabaseCatalog` lookups that's scoped per connection (ownership and lifecycle). This is designed to be injected as a dependency of `CatalogAccessor` at its instantiation, and components requesting information from the `CatalogAccessor` will transparently look in a cache first if it exists. If the cache is passed in as `nullptr`, then the `CatalogAccessor` performs its lookup from the `DatabaseCatalog` as normal. Most operations are expected to only be performed by `CatalogAccessor`, which is why most of this class is private and the `CatalogAccessor` is designated as a friend class.
> 
> Right now it only caches lookups for table and index pointers, and indexes on a table. In the future if we see other `DatabaseCatalog` requests proving expensive, we can add them to the cache.

If this is a design that we're satisfied with, we may consider going back and modifying the `StatementCache` to follow the similar timestamp semantics so it's also resilient to DDL changes.

### CODE REVIEW

File: src/include/catalog/catalog_cache.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    TERRIER_ASSERT(pointers_.count(key) == 0, "Shouldn't be inserting something that already exists.");
+    pointers_[key] = value;
+  }
+
+  std::pair<bool, std::vector<index_oid_t>> GetIndexOids(const table_oid_t table) {
+    const auto it = indexes_.find(table);
+    if (it != indexes_.end()) {
+      return {true, it->second};
+    }
+    return {false, {}};
```
Comment (thepinetree) [inline src/include/catalog/catalog_cache.h:89]: Can you add a comment that the bool is added to differentiate from a table with no indexes

File: src/execution/compiler/codegen.cpp [hunk 2]
```diff
@@ -8,6 +8,7 @@
 #include "date/date.h"
 #include "execution/sql/value.h"
 #include "parser/expression/constant_value_expression.h"
+#include "storage/index/index_defs.h"
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply src/execution/compiler/codegen.cpp:11]: Removed from the higher level catalog_accessor.h, so we need it in more lower-level files now.

File: src/binder/bind_node_visitor.cpp [hunk 3]
```diff
@@ -13,6 +13,7 @@
 #include "catalog/catalog_defs.h"
 #include "common/exception.h"
 #include "common/managed_pointer.h"
+#include "execution/functions/function_context.h"
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply src/binder/bind_node_visitor.cpp:16]: Removed from the higher level catalog_accessor.h, so we need it in more lower-level files now.

File: src/execution/compiler/operator/delete_translator.cpp [hunk 4]
```diff
@@ -5,6 +5,7 @@
 
 #include "execution/compiler/function_builder.h"
 #include "execution/compiler/translator_factory.h"
+#include "storage/index/index.h"
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply src/execution/compiler/operator/delete_translator.cpp:8]: Removed from the higher level catalog_accessor.h, so we need it in more lower-level files now.

File: src/execution/compiler/operator/index_join_translator.cpp [hunk 5]
```diff
@@ -1,9 +1,12 @@
 #include "execution/compiler/operator/index_join_translator.h"
+
 #include <memory>
+
 #include "execution/compiler/function_builder.h"
 #include "execution/compiler/operator/operator_translator.h"
 #include "execution/compiler/translator_factory.h"
 #include "planner/plannodes/index_join_plan_node.h"
+#include "storage/index/index.h"
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply src/execution/compiler/operator/index_join_translator.cpp:9]: Removed from the higher level catalog_accessor.h, so we need it in more lower-level files now.

File: src/execution/sql/storage_interface.cpp [hunk 6]
```diff
@@ -6,6 +6,7 @@
 #include "catalog/catalog_accessor.h"
 #include "execution/exec/execution_context.h"
 #include "execution/util/execution_common.h"
+#include "storage/index/index.h"
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply src/execution/sql/storage_interface.cpp:9]: Removed from the higher level catalog_accessor.h, so we need it in more lower-level files now.

File: src/execution/compiler/operator/index_scan_translator.cpp [hunk 7]
```diff
@@ -1,10 +1,13 @@
 #include "execution/compiler/operator/index_scan_translator.h"
+
 #include <memory>
 #include <unordered_map>
+
 #include "execution/compiler/function_builder.h"
 #include "execution/compiler/operator/operator_translator.h"
 #include "execution/compiler/translator_factory.h"
 #include "planner/plannodes/index_join_plan_node.h"
+#include "storage/index/index.h"
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply src/execution/compiler/operator/index_scan_translator.cpp:10]: Removed from the higher level catalog_accessor.h, so we need it in more lower-level files now.

File: src/execution/compiler/operator/insert_translator.cpp [hunk 8]
```diff
@@ -5,6 +5,7 @@
 
 #include "execution/compiler/function_builder.h"
 #include "execution/compiler/translator_factory.h"
+#include "storage/index/index.h"
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply src/execution/compiler/operator/insert_translator.cpp:8]: Removed from the higher level catalog_accessor.h, so we need it in more lower-level files now.

File: src/execution/compiler/operator/update_translator.cpp [hunk 9]
```diff
@@ -5,6 +5,7 @@
 
 #include "execution/compiler/function_builder.h"
 #include "execution/compiler/translator_factory.h"
+#include "storage/index/index.h"
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply src/execution/compiler/operator/update_translator.cpp:8]: Removed from the higher level catalog_accessor.h, so we need it in more lower-level files now.

File: test/catalog/catalog_test.cpp [hunk 10]
```diff
@@ -8,7 +8,9 @@
 
 #include "catalog/catalog_accessor.h"
 #include "catalog/catalog_defs.h"
+#include "catalog/database_catalog.h"
 #include "catalog/postgres/pg_namespace.h"
+#include "execution/functions/function_context.h"
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply test/catalog/catalog_test.cpp:13]: Removed from the higher level catalog_accessor.h, so we need it in more lower-level files now.

File: test/test_util/sql_table_test_util.cpp [hunk 11]
```diff
@@ -3,6 +3,7 @@
 #include <utility>
 #include <vector>
 
+#include "catalog/database_catalog.h"
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply test/test_util/sql_table_test_util.cpp:6]: Removed from the higher level catalog_accessor.h, so we need it in more lower-level files now.

File: src/traffic_cop/traffic_cop.cpp [hunk 12]
```diff
@@ -367,7 +368,7 @@ std::pair<catalog::db_oid_t, catalog::namespace_oid_t> TrafficCop::CreateTempNam
   }
 
   const auto ns_oid =
-      catalog_->GetAccessor(common::ManagedPointer(txn), db_oid)
+      catalog_->GetAccessor(common::ManagedPointer(txn), db_oid, DISABLED)
```
[the remark this answers is not in the document]
Comment (mbutrovich) [reply src/traffic_cop/traffic_cop.cpp:371]: I don't think the codepath benefits from the lookups that the cache currently offers, and it's rare enough (client disconnecting) that it's not a critical path I really care about speeding up.

**Comment by mbutrovich** [comment]
> From what I saw the catalog cache was always set as a nullptr in the benchmarks. Should there be a test that runs the cache code to ensure no future changes fail in the cache logic, perhaps in catalog_benchmark.cpp?

You need to go over the wire (have a `ConnectionContext`) to exercise this code path, which we don't do in our C++ benchmarks. We rely on JUnit and oltpbench for end-to-end correctness.

**Pull request merged**